File: haskell/p11.hs
-- P11 (*) Modified run-length encoding.

-- From P09.
pack :: Eq x => [x] -> [[x]]
pack []     = []
pack (x:xs) = (x:y) : pack ys
  where (y, ys) = span (== x) xs

data ElemOrTuple x
  = Elem x
  | Tuple Int x deriving (Eq, Show)

myModify :: Eq x => [x] -> [ElemOrTuple x]
myModify x = map (\e -> elemOrTuple e) (pack x)

elemOrTuple :: [x] -> ElemOrTuple x
elemOrTuple (x:[]) = Elem x
elemOrTuple x      = Tuple (length x) (head x)

-- With Either
myModifyEither :: Eq x => [x] -> [Either x (Int,x)]
myModifyEither x = map (\e -> elemOrTupleEither e) (pack x)

elemOrTupleEither :: [x] -> Either x (Int,x)
elemOrTupleEither (x:[]) = Left x
elemOrTupleEither x      = Right (length x, head x)

File: ben/prolog/P01-P28-lists/p11.pl
% Modified run-length encoding.

pack_first([], _, [], []). 
pack_first([X|Xs], X, R, [X|Ys]) :-
    pack_first(Xs, X, R, Ys).
pack_first([X|Xs], Y, [X|Xs], []) :- 
    X \= Y.

encode_modified([], []).
encode_modified([Y|Ys], [Y |Xs]) :-
    pack_first([Y|Ys], Y, L, X),
    length(X, 1),
    encode_modified(L, Xs).
encode_modified([Y|Ys], [[Len, Y] |Xs]) :-
    pack_first([Y|Ys], Y, L, X),
    length(X, Len),
    Len \= 1,
    encode_modified(L, Xs).
